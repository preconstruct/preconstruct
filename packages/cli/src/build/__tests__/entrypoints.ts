import build from "../";
import path from "path";
import {
  snapshotDirectory,
  stripHashes,
  testdir,
  js,
  getFiles,
} from "../../../test-utils";
import spawn from "spawndamnit";

jest.setTimeout(10000);

jest.mock("../../prompt");

test("multiple entrypoints", async () => {
  let dir = await testdir({
    "package.json": JSON.stringify({
      name: "multiple-entrypoints",
      main: "dist/multiple-entrypoints.cjs.js",
      module: "dist/multiple-entrypoints.esm.js",
      preconstruct: {
        entrypoints: ["index.js", "multiply.js"],
      },
    }),
    "multiply/package.json": JSON.stringify({
      main: "dist/multiple-entrypoints-multiply.cjs.js",
      module: "dist/multiple-entrypoints-multiply.esm.js",
    }),
    "src/index.js": js`
      export let sum = (a, b) => a + b;
    `,
    "src/multiply.js": js`
      export let multiply = (a, b) => a * b;
    `,
  });

  await build(dir);

  expect(await getFiles(dir, ["**/dist/**"])).toMatchInlineSnapshot(`
    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ dist/multiple-entrypoints.cjs.dev.js, dist/multiple-entrypoints.cjs.prod.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    'use strict';

    Object.defineProperty(exports, '__esModule', { value: true });

    let sum = (a, b) => a + b;

    exports.sum = sum;

    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ dist/multiple-entrypoints.cjs.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    'use strict';

    if (process.env.NODE_ENV === "production") {
      module.exports = require("./multiple-entrypoints.cjs.prod.js");
    } else {
      module.exports = require("./multiple-entrypoints.cjs.dev.js");
    }

    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ dist/multiple-entrypoints.esm.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    let sum = (a, b) => a + b;

    export { sum };

    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ multiply/dist/multiple-entrypoints-multiply.cjs.dev.js, multiply/dist/multiple-entrypoints-multiply.cjs.prod.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    'use strict';

    Object.defineProperty(exports, '__esModule', { value: true });

    let multiply = (a, b) => a * b;

    exports.multiply = multiply;

    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ multiply/dist/multiple-entrypoints-multiply.cjs.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    'use strict';

    if (process.env.NODE_ENV === "production") {
      module.exports = require("./multiple-entrypoints-multiply.cjs.prod.js");
    } else {
      module.exports = require("./multiple-entrypoints-multiply.cjs.dev.js");
    }

    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ multiply/dist/multiple-entrypoints-multiply.esm.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    let multiply = (a, b) => a * b;

    export { multiply };

  `);
});

test("two entrypoints, one module, one not", async () => {
  let dir = await testdir({
    "package.json": JSON.stringify({
      name: "two-entrypoints-one-module-one-not",
      main: "dist/two-entrypoints-one-module-one-not.cjs.js",
      preconstruct: {
        entrypoints: ["index.js", "multiply.js"],
      },
    }),
    "multiply/package.json": JSON.stringify({
      main: "dist/two-entrypoints-one-module-one-not-multiply.cjs.js",
      module: "dist/two-entrypoints-one-module-one-not-multiply.esm.js",
    }),
    "src/index.js": js`
      export let sum = (a, b) => a + b;
    `,
    "src/multiply.js": js`
      export let multiply = (a, b) => a * b;
    `,
  });

  await expect(build(dir)).rejects.toMatchInlineSnapshot(
    `[Error: two-entrypoints-one-module-one-not/multiply has a module build but two-entrypoints-one-module-one-not does not have a module build. Entrypoints in a package must either all have a particular build type or all not have a particular build type.]`
  );
});

test("two entrypoints with a common dependency", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "common-dependency-two-entrypoints",
      main: "dist/common-dependency-two-entrypoints.cjs.js",
      module: "dist/common-dependency-two-entrypoints.esm.js",

      preconstruct: {
        entrypoints: ["index.js", "multiply.js"],
      },
    }),

    "multiply/package.json": JSON.stringify({
      main: "dist/common-dependency-two-entrypoints-multiply.cjs.js",
      module: "dist/common-dependency-two-entrypoints-multiply.esm.js",
    }),

    "src/identity.js": js`
      export let identity = (x) => x;
    `,

    "src/multiply.js": js`
      import { identity } from "./identity";

      export let multiply = (a, b) => identity(a * b);

      export { identity };
    `,

    "src/index.js": js`
      import { identity } from "./identity";

      export let sum = (a, b) => identity(a + b);

      export { identity };
    `,
  });

  await build(tmpPath);

  await snapshotDirectory(tmpPath, await stripHashes("identity"));
  let root = require(tmpPath);
  let other = require(path.join(tmpPath, "multiply"));
  expect(typeof root.identity).toBe("function");
  expect(root.identity).toBe(other.identity);
});

test("two entrypoints where one requires the other entrypoint", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "importing-another-entrypoint",
      main: "dist/importing-another-entrypoint.cjs.js",

      preconstruct: {
        entrypoints: ["index.js", "multiply.js"],
      },
    }),

    "multiply/package.json": JSON.stringify({
      main: "dist/importing-another-entrypoint-multiply.cjs.js",
    }),

    "src/index.js": js`
      export let identity = (x) => x;
    `,

    "src/multiply.js": js`
      import { identity } from "./index";

      export let multiply = (a, b) => identity(a * b);
    `,
  });

  await build(tmpPath);

  await snapshotDirectory(tmpPath);

  let { identity } = require(tmpPath);
  expect(identity(20)).toBe(20);

  let { multiply } = require(path.join(tmpPath, "multiply"));

  expect(multiply(2, 3)).toBe(6);
});

test("multiple entrypoints", async () => {
  let dir = await testdir({
    "package.json": JSON.stringify({
      name: "multiple-entrypoints",
      main: "dist/multiple-entrypoints.cjs.js",
      module: "dist/multiple-entrypoints.esm.js",
      exports: {
        ".": {
          types: "./dist/multiple-entrypoints.cjs.js",
          module: "./dist/multiple-entrypoints.esm.js",
          default: "./dist/multiple-entrypoints.cjs.js",
        },
        "./multiply": {
          types: "./dist/multiple-entrypoints-multiply.cjs.js",
          module: "./dist/multiple-entrypoints-multiply.esm.js",
          default: "./dist/multiple-entrypoints-multiply.cjs.js",
        },
        "./package.json": "./package.json",
      },
      preconstruct: {
        exports: true,
        entrypoints: ["index.js", "multiply.js"],
        ___experimentalFlags_WILL_CHANGE_IN_PATCH: {
          importsConditions: true,
          distInRoot: true,
        },
      },
    }),
    "multiply/package.json": JSON.stringify({
      main: "../dist/multiple-entrypoints-multiply.cjs.js",
      module: "../dist/multiple-entrypoints-multiply.esm.js",
    }),
    "src/index.js": js`
      export let sum = (a, b) => a + b;
    `,
    "src/multiply.js": js`
      export let multiply = (a, b) => a * b;
    `,
    "runtime-blah.js": js`
      const { multiply } = require("multiple-entrypoints/multiply");
      console.log(multiply(2, 2) + "");
    `,
  });

  await build(dir);

  expect(await getFiles(dir, ["**/dist/**"])).toMatchInlineSnapshot(`
    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ dist/multiple-entrypoints-multiply.cjs.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    'use strict';

    Object.defineProperty(exports, '__esModule', { value: true });

    let multiply = (a, b) => a * b;

    exports.multiply = multiply;

    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ dist/multiple-entrypoints-multiply.esm.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    let multiply = (a, b) => a * b;

    export { multiply };

    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ dist/multiple-entrypoints.cjs.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    'use strict';

    Object.defineProperty(exports, '__esModule', { value: true });

    let sum = (a, b) => a + b;

    exports.sum = sum;

    ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ dist/multiple-entrypoints.esm.js ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
    let sum = (a, b) => a + b;

    export { sum };

  `);
  let node = await spawn("node", ["runtime-blah.js"], { cwd: dir });

  expect(node.stdout.toString("utf8")).toMatchInlineSnapshot(`
    "4
    "
  `);
  expect(node.stderr.toString("utf8")).toMatchInlineSnapshot(`""`);
  expect(node.code).toBe(0);
});
