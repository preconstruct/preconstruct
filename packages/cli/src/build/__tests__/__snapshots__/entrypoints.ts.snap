// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`two entrypoints where one requires the other entrypoint: dist/importing-another-entrypoint.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

let identity = x => x;

exports.identity = identity;
"
`;

exports[`two entrypoints where one requires the other entrypoint: dist/importing-another-entrypoint.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === "production") {
  module.exports = require("./importing-another-entrypoint.cjs.prod.js");
} else {
  module.exports = require("./importing-another-entrypoint.cjs.dev.js");
}
"
`;

exports[`two entrypoints where one requires the other entrypoint: dist/importing-another-entrypoint.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

let identity = x => x;

exports.identity = identity;
"
`;

exports[`two entrypoints where one requires the other entrypoint: multiply/dist/importing-another-entrypoint-multiply.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dist_importingAnotherEntrypoint = require('../../dist/importing-another-entrypoint.cjs.dev.js');

let multiply = (a, b) => dist_importingAnotherEntrypoint.identity(a * b);

exports.multiply = multiply;
"
`;

exports[`two entrypoints where one requires the other entrypoint: multiply/dist/importing-another-entrypoint-multiply.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === "production") {
  module.exports = require("./importing-another-entrypoint-multiply.cjs.prod.js");
} else {
  module.exports = require("./importing-another-entrypoint-multiply.cjs.dev.js");
}
"
`;

exports[`two entrypoints where one requires the other entrypoint: multiply/dist/importing-another-entrypoint-multiply.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dist_importingAnotherEntrypoint = require('../../dist/importing-another-entrypoint.cjs.prod.js');

let multiply = (a, b) => dist_importingAnotherEntrypoint.identity(a * b);

exports.multiply = multiply;
"
`;

exports[`two entrypoints where one requires the other entrypoint: src/index.js 1`] = `"export let identity = (x) => x;"`;

exports[`two entrypoints where one requires the other entrypoint: src/multiply.js 1`] = `
"import { identity } from "./index";

export let multiply = (a, b) => identity(a * b);"
`;

exports[`two entrypoints with a common dependency: dist/common-dependency-two-entrypoints.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var identity = require('./identity-some-hash.cjs.dev.js');

let sum = (a, b) => identity.identity(a + b);

exports.identity = identity.identity;
exports.sum = sum;
"
`;

exports[`two entrypoints with a common dependency: dist/common-dependency-two-entrypoints.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === "production") {
  module.exports = require("./common-dependency-two-entrypoints.cjs.prod.js");
} else {
  module.exports = require("./common-dependency-two-entrypoints.cjs.dev.js");
}
"
`;

exports[`two entrypoints with a common dependency: dist/common-dependency-two-entrypoints.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var identity = require('./identity-some-hash.cjs.prod.js');

let sum = (a, b) => identity.identity(a + b);

exports.identity = identity.identity;
exports.sum = sum;
"
`;

exports[`two entrypoints with a common dependency: dist/common-dependency-two-entrypoints.esm.js 1`] = `
"import { i as identity } from './identity-some-hash.esm.js';
export { i as identity } from './identity-some-hash.esm.js';

let sum = (a, b) => identity(a + b);

export { sum };
"
`;

exports[`two entrypoints with a common dependency: dist/identity-this-is-not-the-real-hash-09d8aa2cd41dd37e55ea1ca5c8546d35.esm.js 1`] = `
"let identity = x => x;

export { identity as i };
"
`;

exports[`two entrypoints with a common dependency: dist/identity-this-is-not-the-real-hash-f57161fefd97e3101601d153708a46b8.cjs.dev.js 1`] = `
"'use strict';

let identity = x => x;

exports.identity = identity;
"
`;

exports[`two entrypoints with a common dependency: dist/identity-this-is-not-the-real-hash-f57161fefd97e3101601d153708a46b8.cjs.prod.js 1`] = `
"'use strict';

let identity = x => x;

exports.identity = identity;
"
`;

exports[`two entrypoints with a common dependency: multiply/dist/common-dependency-two-entrypoints-multiply.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var identity = require('../../dist/identity-some-hash.cjs.dev.js');

let multiply = (a, b) => identity.identity(a * b);

exports.identity = identity.identity;
exports.multiply = multiply;
"
`;

exports[`two entrypoints with a common dependency: multiply/dist/common-dependency-two-entrypoints-multiply.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === "production") {
  module.exports = require("./common-dependency-two-entrypoints-multiply.cjs.prod.js");
} else {
  module.exports = require("./common-dependency-two-entrypoints-multiply.cjs.dev.js");
}
"
`;

exports[`two entrypoints with a common dependency: multiply/dist/common-dependency-two-entrypoints-multiply.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var identity = require('../../dist/identity-some-hash.cjs.prod.js');

let multiply = (a, b) => identity.identity(a * b);

exports.identity = identity.identity;
exports.multiply = multiply;
"
`;

exports[`two entrypoints with a common dependency: multiply/dist/common-dependency-two-entrypoints-multiply.esm.js 1`] = `
"import { i as identity } from '../../dist/identity-some-hash.esm.js';
export { i as identity } from '../../dist/identity-some-hash.esm.js';

let multiply = (a, b) => identity(a * b);

export { multiply };
"
`;

exports[`two entrypoints with a common dependency: src/identity.js 1`] = `"export let identity = (x) => x;"`;

exports[`two entrypoints with a common dependency: src/index.js 1`] = `
"import { identity } from "./identity";

export let sum = (a, b) => identity(a + b);

export { identity };"
`;

exports[`two entrypoints with a common dependency: src/multiply.js 1`] = `
"import { identity } from "./identity";

export let multiply = (a, b) => identity(a * b);

export { identity };"
`;
